### PR TITLE
Activerecord where exists relation

### PR DESCRIPTION
**Pull request by urusaich**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | ? |

Allows to use relations in exists where.

``` php
$document->andWhere(['exists', 'comments']);
```

result:

``` sql
SELECT * FROM {{%document}} WHERE EXISTS (SELECT * FROM {{%comment}} WHERE {{%document}}.[[id]] = {{%comment}}.[[doc_id]])
```

``` php
$document->andWhere(['and', ['exists', 'comments'], ['exists', 'opens']]);
```

result:

``` sql
SELECT * FROM {{%document}} WHERE (EXISTS (SELECT * FROM {{%comment}} WHERE {{%document}}.[[id]] = {{%comment}}.[[doc_id]])) AND (EXISTS (SELECT * FROM {{%open}} WHERE {{%document}}.[[id]] = {{%open}}.[[doc_id]]))
```

``` php
$document->andWhere(['exists', 'comments' => function(ActiveQuery $q){
    $q->alias('alias');
}]);
```

result:

``` sql
SELECT * FROM {{%document}} WHERE EXISTS (SELECT * FROM {{%comment}} `alias` WHERE {{%document}}.[[id]] = {{alias}}.[[doc_id]])
```

### CODE REVIEW

**Comment by samdark** [comment]
@yiisoft/core-developers thoughts?

**Comment by SilverFire** [comment]
Looks interesting and useful

File: framework/db/ActiveQuery.php [hunk 1]
```diff
@@ -159,6 +159,10 @@ public function prepare($builder)
             $this->select = ["$alias.*"];
         }
 
+        if (!empty($this->where)) {
+            $this->where = $this->buildExistsRelations($this->where);
```
Comment (SilverFire) [inline framework/db/ActiveQuery.php]: Is is really necessary to call `buildExistsRelations()` in each `prepare()`?

File: framework/db/ActiveQuery.php [hunk 2]
```diff
@@ -728,6 +732,78 @@ public function orOnCondition($condition, $params = [])
...
+     */
+    private function buildExistsRelations($where)
+    {
+        if (!isset($where[0]) || !is_array($where)) {
+            return $where;
+        }
+
+        $operator = array_shift($where);
+        /** @var ActiveRecord $model */
+        $model = new $this->modelClass;
```
Comment (SilverFire) [inline framework/db/ActiveQuery.php]: It is not OK to create new `ActiveRecord` object on each call even when method has nothing to do

Comment (nanodesu88) [inline framework/db/ActiveQuery.php]: True. Whats about replace `new $this->modelClass` with `Yii::createObject($this->modelClass)`?

Comment (nanodesu88) [inline framework/db/ActiveQuery.php]: True. Modified.

File: framework/db/ActiveQuery.php [hunk 3]
```diff
@@ -728,6 +732,88 @@ public function orOnCondition($condition, $params = [])
     }
 
     /**
+     * Starts a transform relations in exists condition
+     */
+    private function buildExistsRelations()
```
Comment (cebe) [inline framework/db/ActiveQuery.php]: this  should be implemented in QueryBuilder instead.

Comment (nanodesu88) [inline framework/db/ActiveQuery.php]: Why? Is QueryBuilder knows something about Active db components?
`buildJoinWith`, `joinWithRelations` and `joinWithRelation` builds join from relations. 
`buildExistsRelations`, `existsRelations` and `existsRelation` does the same, but builds exists where instead.

**Comment by samdark** [comment]
Won't make it into 2.0.

**Pull request closed without merging**